### PR TITLE
Library Redesign

### PR DESCRIPTION
**Pull request by puniththota**
Library Redesign

### CODE REVIEW

File: Test/JsonDiffTest.cs [hunk 1]
```diff
@@ -66,22 +66,28 @@ public static IEnumerable<object[]> ProviderForNumericDiffTest()
...
+                                                      'Division': 1
+                                                      }
                                                 }
                                                 },
                                     '1': {
                                           'c': {
                                                    'left': 25,
                                                    'right': 20,
-                                                    'diff': 5
+                                                    'Int Diff': {
```
Comment (ColmBhandal) [inline Test/JsonDiffTest.cs]: This doesn't really test much different behaviour or edge cases to the (10, 15) case. To vary the tests as much as possible, reverse the order of the 20 and the 25 and confirm that the subtraction diff comes out at -1. I think the division would stay the same

File: c_sharp_json_diff/DiffProcessor.cs [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace c_sharp_json_diff
+{
+    abstract class DiffProcessor : IJsonProcessor<object>
+    {
+        private const string SkippableProperty = "_t";
+        private const string propertyLeft = "left";
```
Comment (ColmBhandal) [inline c_sharp_json_diff/DiffProcessor.cs]: Rename to ``keyLeft`` and ``keyRight``

File: Test/JsonDiffTest.cs [hunk 3]
```diff
@@ -66,22 +66,28 @@ public static IEnumerable<object[]> ProviderForNumericDiffTest()
                                      { 'b': 10 },
                                      { 'c': 20 },
                                     ]
-                                 }", 
+                                 }",
                                @"{
                                'a': {
```
Comment (ColmBhandal) [inline Test/JsonDiffTest.cs]: Rename 'a' so that it's not confused with the array type in '_t': 'a'

File: Test/JsonDiffTest.cs [hunk 4]
```diff
@@ -66,22 +66,28 @@ public static IEnumerable<object[]> ProviderForNumericDiffTest()
...
                                     '_t': 'a',
                                      '0': {
                                            'b': {
                                                    'left': 15,
                                                    'right': 10,
-                                                    'diff': 5
+                                                    'Int Diff': {
+                                                      'Subtraction': 5,
```
Comment (ColmBhandal) [inline Test/JsonDiffTest.cs]: The usual convention is that subtraction is right minus left, not left minus right. I would change this

File: c_sharp_json_diff/JsonDiff.cs [hunk 5]
```diff
@@ -67,71 +70,17 @@ public string NumericDiffAsString(string left, string right)
...
-            return jObject;
-        }
-
         private string JsonDiffBetweenTwo(string left, string right)
         {
             return new JsonDiffPatch(new Options { ArrayDiff = ArrayDiffMode.Efficient }).Diff(left, right);
         }
 
-        private Dictionary<object, object> CreateDictionary(object left, object right, object diff)
+        private void InjectProcessors(Dictionary<string, IJsonProcessor<object>> processors, JObject jObject)
```
Comment (ColmBhandal) [inline c_sharp_json_diff/JsonDiff.cs]: The order of the processors is important. We expect processors to mutate the ``jObject`` . So we should pass the processors as a list rather than a dictionary. We are only using the value anyway, so a dictionary is not needed here.

File: c_sharp_json_diff/IntDiffProcessor.cs [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+        public override Dictionary<string, object> PerformDiffProcess(JToken left, JToken right)
+        {
+            string leftValueAsString = left.ToString();
+            string rightValueAsString = right.ToString();
+            NumericDiff diff = new NumericDiff();
+
+            if (int.TryParse(leftValueAsString, out int leftValueAsInt) && int.TryParse(rightValueAsString, out int rightValueAsInt))
+            {
+                diff.Subtraction = leftValueAsInt - rightValueAsInt;
```
Comment (ColmBhandal) [inline c_sharp_json_diff/IntDiffProcessor.cs]: Change the order to right minus left and right / left. This is the usual way that deltas/diffs are calculated.

File: c_sharp_json_diff/DoubleDiffProcessor.cs [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+            string rightValueAsString = right.ToString();
+            Dictionary<string, object> dict = new Dictionary<string, object>();
+
+            if (!IsIntable(leftValueAsString, rightValueAsString))
+            {
+                NumericDiff diff = new NumericDiff();
+
+                if (double.TryParse(leftValueAsString, out double leftValueAsDouble) && double.TryParse(rightValueAsString, out double rightValueAsDouble))
+                {
+                    diff.Subtraction = leftValueAsDouble - rightValueAsDouble;
```
Comment (ColmBhandal) [inline c_sharp_json_diff/DoubleDiffProcessor.cs]: Change the order to right minus left and right / left. This is the usual way that deltas/diffs are calculated.

File: c_sharp_json_diff/DoubleDiffProcessor.cs [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+                {
+                    dict.Add("Subtraction", diff.Subtraction);
+                    dict.Add("Division", diff.Division);
+                }
+            }
+
+            return dict;
+        }
+
+        private bool IsIntable(string left, string right)
```
Comment (ColmBhandal) [inline c_sharp_json_diff/DoubleDiffProcessor.cs]: Actually, I think we need to revisit this. The tests have made it clearer. Testing a try-parse on the two values seems to be too strict. For example, if the left is 20 and the right is 25, then the ratio right/left = 25/20 which is 5/4 or 1.25. This is clearly a double value, even though the two inputs are ints. 

So I would say we need to make ``IsIntable`` stricter, and probably rename it to something like ``IsIntDiffable``. We should still leave in the test that both left and right are ints. But to make it stricter, we should also confirm that ``right % left == 0``. That is, the remainder after division should be zero, meaning one divides perfectly into the other.

File: c_sharp_json_diff/JsonDiff.cs [hunk 9]
```diff
@@ -45,8 +45,11 @@ public JObject NumericDiff(string left, string right)
                     return JObject.Parse(@"{}");
                 }
                 JObject diffJsonObject = JObject.Parse(diff);
-                JObject diffReturnObject = CalculateNumericDiff(diffJsonObject, _isAbsoluteDiff);
-                return diffReturnObject;
+                Dictionary<string, IJsonProcessor<object>> processors = new Dictionary<string, IJsonProcessor<object>>();
```
Comment (ColmBhandal) [inline c_sharp_json_diff/JsonDiff.cs]: Processors should be refactored to a list (also added comment below)

File: c_sharp_json_diff/DiffProcessor.cs [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+                        if(keyValuePair.Value is JObject jsonObj)
+                        {
+                            Process(key, jsonObj);
+                        }
+                    }
+                }
+            }
+        }
+
+        public abstract Dictionary<string, object> PerformDiffProcess(JToken left, JToken right);
```
Comment (ColmBhandal) [inline c_sharp_json_diff/DiffProcessor.cs:53]: Add comment-documentation here about what this dictionary represents. Include in the documentation that an empty or null dictionary means we skip the diff.

File: c_sharp_json_diff/DoubleDiffProcessor.cs [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+{
+    class DoubleDiffProcessor : DiffProcessor
+    {
+        public override Dictionary<string, object> PerformDiffProcess(JToken left, JToken right)
+        {
+            string leftValueAsString = left.ToString();
+            string rightValueAsString = right.ToString();
+            Dictionary<string, object> dict = new Dictionary<string, object>();
+
+            if (!IsIntable(leftValueAsString, rightValueAsString))
```
Comment (ColmBhandal) [inline c_sharp_json_diff/DoubleDiffProcessor.cs]: Add a property to the class, that controls how this class processes ints. For example, it could be called ``IntProcessing``. We could make it an enum with these possible values:

1. IgnoreInts - If both left and right are ints, the processor skips that node
2. ProcessIfFraction - It will process left and right if they are both ints but if left does not perfectly divide right
3. ProcessAll - It will process left and right if they are ints, regardless

File: c_sharp_json_diff/IntDiffProcessor.cs [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+    class IntDiffProcessor : DiffProcessor
+    {
+
+        public override Dictionary<string, object> PerformDiffProcess(JToken left, JToken right)
+        {
+            string leftValueAsString = left.ToString();
+            string rightValueAsString = right.ToString();
+            NumericDiff diff = new NumericDiff();
+
+            if (int.TryParse(leftValueAsString, out int leftValueAsInt) && int.TryParse(rightValueAsString, out int rightValueAsInt))
```
Comment (ColmBhandal) [inline c_sharp_json_diff/IntDiffProcessor.cs]: Add property to this similar to the ``DoubleDiffProcessor``'s ``IntProcessing`` property I proposed in another comment. This could have two values:

1. PerfectDivisionOnly - It will only process if both are ints and left perfectly divides into right
2. BothInts - It will process if both are ints

File: c_sharp_json_diff/JsonDiff.cs [hunk 13]
```diff
@@ -45,8 +45,11 @@ public JObject NumericDiff(string left, string right)
...
+                processors.Add("Int Diff", new IntDiffProcessor());
```
Comment (ColmBhandal) [inline c_sharp_json_diff/JsonDiff.cs]: Set the ``IntProcessing`` property (proposed in another comment) on the Int/Double processors to ensure that they are processing sensibly. 

For example, an int processor should not fire if the input values are 20 and 25 for left and right because the result would be the fraction 1.25.

For this to work, you may also need to re-order the processors in the list so that the double processor comes first.

**Review by ColmBhandal — CHANGES_REQUESTED** [review]
Good progress. Main thing that I thin needs to change is that double processors should kick in whenever int division is imperfect, for the numeric diff. For that, I propose adding properties to control the strength of matching on the different diff classes.

Another thing that needs to change is the order of the diff - in most applications a diff is right minus left or right divided by left.

Inline are more detailed comments explaining this, and suggesting some other minor tweaks.

**Comment by ColmBhandal** [comment]
We should also have ``DoubleDiff`` as a top-level function. For our use case, it will be impractical to have to look for both IntDiff and DoubleDiff labels. If we can just run a DoubleDiff then all we'll need to look for are the DoubleDiff labels. In this case, the DoubleDiff top-level function should also take a parameter to control how strict the diff is - match all ints, match only strict doubles etc. This would correspond to other comments I've added about a property on the DoubleDiff class.

The same goes for ``IntDiff`` - a top level function should be exposed along with a way to control different levels of strictness.

File: c_sharp_json_diff/JsonDiff.cs (hunk 13 above)
Comment (hctrdev) [reply c_sharp_json_diff/JsonDiff.cs]: What is the value in having an int processor at all? An Int can always be cast to a Double, so why not ignore Ints and just deal with doubles? The client can handle rounding the values to Ints if they need to.

**Review by hctrdev — CHANGES_REQUESTED** [review]
I agree with all of Colm's comments.

But I am still a bit confused about why we need both a double diff and an int diff. To me, they are pretty much the same thing and the double diff alone is enough. I don't see the value in using a different label for the diff object if both inputs happen to be ints. It seems to just make everything more complex.

File: c_sharp_json_diff/JsonDiff.cs (hunk 13 above)
Comment (ColmBhandal) [reply c_sharp_json_diff/JsonDiff.cs]: Good point. I agree that there isn't much use in the Int Diff processor as a Diff processor. If we want Ints, we can just post-process the output of the Double processor using a DoubleToInt diff conversion processor. But we don't need that right now, so it would be fine to just get rid of the Int processor altogether for now.

@puniththota I am actually swaying towards Hector's advice: just get rid of the Int Diff processor & instead of exposing a "numeric diff" function, just expose a "Double Diff" function. Then a large swath of my comments will no longer be relevant and it'll be a lot less work for you.

File: c_sharp_json_diff/JsonDiff.cs (hunk 5 above)
Comment (puniththota) [reply c_sharp_json_diff/JsonDiff.cs]: Refactored to be list

File: c_sharp_json_diff/DoubleDiffProcessor.cs (hunk 8 above)
Comment (puniththota) [reply c_sharp_json_diff/DoubleDiffProcessor.cs]: As discussed, removed int processor, so this is no longer required

File: c_sharp_json_diff/IntDiffProcessor.cs (hunk 12 above)
Comment (puniththota) [reply c_sharp_json_diff/IntDiffProcessor.cs]: As discussed, removed int processor, so this is no longer required

**Review by ColmBhandal — APPROVED** [review]

**Review by hctrdev — APPROVED** [review]
